compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp

```diff
@@ -174,6 +176,225 @@ void transform_dialect::ApplyFoldFillIntoPadPatternsOp::populatePatterns(
   patterns.insert<FoldFillIntoPad>(patterns.getContext());
 }
 
+//===---------------------------------------------------------------------===//
+// ApplyHoistForallFromForPatternsOp
+//===---------------------------------------------------------------------===//
+
+namespace {
+struct HoistForallFromFor : public OpRewritePattern<scf::ForOp> {
+  using OpRewritePattern<scf::ForOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(scf::ForOp loop,
+                                PatternRewriter &rewriter) const final {
+    if (loop.getBody()->getOperations().size() != 2) {
+      return rewriter.notifyMatchFailure(
+          loop, "Body of the loop contains more than one op");
+    }
+
+    auto forallOp =
```

Comment:
To restrictive? Maybe just look for a single forall op in the body of the outer for?

Reply:
I wanted to slow walk it a bit. Allowing other operations in the loop means we will be moving "block level code" into a thread level context, which I'd prefer to be prescriptive about.

Reply:
I think there is no difference. Outside of the scf.forall its all thread level code (all threads are doing that redundantly). But I see your point.

Reply:
You're probably right. I'm going to land this as is and address this in a follow up if it becomes an issue then.